scripts/test_screeps_rl_training_runner.py

```
@@ -587,10 +614,216 @@ def test_failed_variant_runs_are_excluded_from_reward_ranking(self) -> None:
+                [baseline, candidate, unexpected],
+                r"unexpected-variant-row .*run_index=0.*unexpected variant id 'intruder'",
+            ),
+        )
+        for report_id, variants, message_regex in cases:
+            with self.subTest(report_id=report_id), tempfile.TemporaryDirectory() as temp_dir:
+                root = Path(temp_dir)
+                card_path = root / "card.json"
+                out_dir = root / "reports"
+
+                class MalformedVariantSimulator:
+                    def __call__(self, **kwargs: Any) -> JsonObject:
```

Ruff (0.15.12)

[warning] Dynamically typed expressions (typing.Any) are disallowed in `**kwargs`

(ANN401)